src/components/DataCollector/validateUserEventOptions.js

```diff
@@ -10,18 +10,39 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 governing permissions and limitations under the License.
 */
 
+import { string, objectOf, boolean, arrayOf } from "../../utils/validation";
+
 /**
  * Verifies user provided event options.
  * @param {*} options The user event options to validate
- * @returns {Array} Array of warnings if the options are invalid
+ * @returns {Array} Array of errors if the options are invalid
  */
 export default options => {
-  const warnings = [];
-  const { xdm } = options;
-  if (!xdm && !options.data) {
-    warnings.push("No event xdm or event data specified.");
-  } else if (xdm && !xdm.eventType && !options.type) {
-    warnings.push("No type or xdm.eventType specified.");
+  const eventOptionsValidator = objectOf({
+    viewStart: boolean(),
+    type: string().nonEmpty(),
+    xdm: objectOf({
+      eventType: string().nonEmpty()
+    }).nonEmpty(),
+    data: objectOf({}).nonEmpty(),
+    scopes: arrayOf(
+      string()
+        .required()
+        .nonEmpty()
+    ).nonEmpty()
+  })
+    .required()
+    .nonEmpty();
+  const errors = [];
+  try {
+    const { type, xdm, data, scopes } = eventOptionsValidator(options);
+    if (!xdm && !data && !scopes) {
+      errors.push("No event xdm, data, or scopes specified.");
+    } else if (xdm && !xdm.eventType && !type) {
+      errors.push("No type or xdm.eventType specified.");
+    }
```

Comment:
I see custom validation logic being necessary in other places as well and having to reproduce some of this logic in other components seems a bit redundant. Maybe we could add a validator that allows for custom logic? It could look something like:

```
objectOf({
  viewStart: boolean(),
  type: string().nonEmpty(),
  xdm: objectOf({
    eventType: string().nonEmpty()
  }).nonEmpty(),
  data: objectOf({}).nonEmpty(),
  scopes: arrayOf(
    string()
      .required()
      .nonEmpty()
  ).nonEmpty()
})
  .required()
  .nonEmpty()
  .test(options => {
    const { xdm, data, scopes } = options;
    if (!xdm && !data && !scopes) {
      return "No event xdm, data, or scopes specified.";
    }
  })
  .test(options => {
    const { type, xdm } = options;
    if (xdm && !xdm.eventType && !type) {
      return "No type or xdm.eventType specified.";
    }
  });
```

This would also allow the value of the `validator` property in the systemic approach I proposed in https://jira.corp.adobe.com/browse/CORE-42341 to be an actual validator, rather than having to support a custom validate function. @jonsnyder curious to get your take on this since you're more familiar with the validators.

Reply:
A custom validator matches some of the other frameworks that I've seen and that would probably be a nice addition.

Reply:
Aaron, I like your idea of the test validator.  The current API for validators is to return the modified value, and throw an exception if it is invalid, so I'd rather have that used in a test like this:
```
.test(value => {
  const { xdm, data, scopes } = options;
  if (!xdm && !data && !scopes) {
    throw new Error("No event xdm, data, or scopes specified.");
  }
  return value;
})
```

Reply:
I like it. I've created a JIRA issue: https://jira.corp.adobe.com/browse/CORE-42407. We can address it in a different PR.